packages/devtools_app/lib/src/screens/memory/primitives/class_name.dart

```diff
@@ -0,0 +1,285 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:vm_service/vm_service.dart';
+
+import '../../../primitives/simple_items.dart';
+
+@immutable
+class HeapClassName {
+  HeapClassName({required this.className, required library})
+      : library = _normalizeLibrary(library) {
+    assert(
+      !isCore || !isDartOrFlutter,
+      'isCore and isDartOrFlutter must be exclusive',
+    );
+  }
+
+  HeapClassName.fromClassRef(ClassRef? classRef)
+      : this(
+          library: _library(
+            classRef?.library?.name,
+            classRef?.library?.uri,
+          ),
+          className: classRef?.name ?? '',
+        );
+
+  HeapClassName.fromHeapSnapshotClass(HeapSnapshotClass? theClass)
+      : this(
+          library: _library(
+            theClass?.libraryName,
+            theClass?.libraryUri.toString(),
+          ),
+          className: theClass?.name ?? '',
+        );
+
+  static String _library(String? libName, String? libUrl) {
+    libName ??= '';
+    if (libName.isNotEmpty) return libName;
+    return libUrl ?? '';
+  }
+
+  final String className;
+  final String library;
+
+  String get fullName => library.isNotEmpty ? '$library/$className' : className;
+
+  bool get isSentinel => className == 'Sentinel' && library.isEmpty;
+
+  /// Detects if a class can retain an object from garbage collection.
+  bool get isWeakEntry {
+    // Classes that hold reference to an object without preventing
+    // its collection.
+    const weakHolders = {
+      '_WeakProperty': '${PackagePrefixes.dart}core',
+      '_WeakReferenceImpl': '${PackagePrefixes.dart}core',
+      'FinalizerEntry': '${PackagePrefixes.dart}_internal',
+    };
+
+    if (!weakHolders.containsKey(className)) return false;
+    if (weakHolders[className] == library) return true;
+
+    // If a class lives in unexpected library, this can be because of
+    // (1) name collision or (2) bug in this code.
+    // Throwing exception in debug mode to verify option #2.
+    // TODO(polina-c): create a way for users to add their weak classes
+    // or detect weak references automatically, without hard coding
+    // class names.
+    assert(false, 'Unexpected library for $className: $library.');
+    return false;
+  }
+
+  /// True, if the library is a core library.
+  ///
+  /// I.e. if the library name is empty or does not have prefix
+  /// `dart:` or `package:`.
+  bool get isCore =>
+      library.isEmpty ||
+      (!library.startsWith(PackagePrefixes.dart) &&
+          !library.startsWith(PackagePrefixes.genericDartPackage));
+
+  /// True, if the package has prefix `dart:` or has perfix `package:` and is
+  /// published by Dart or Flutter org.
+  bool get isDartOrFlutter {
+    if (library.startsWith(PackagePrefixes.dart)) return true;
```

Comment:
can we also add a shortcut check for flutter libraries so we don't have to perform List.contains for every Flutter lib?